Add git packages as VCS repositories

mortenscheel opened the pull request:
I've changed the way `packager:git` works.
In stead of cloning the repository directly to the `/packages/myVendor/myPackage` folder, it is added as a VCS repository in `composer.json` and installed like any other package to the `/vendor` folder. After installing, it is symlinked to the `/packages` folder.

If the package origin is only specified as `vendor/package` in stead of a complete URL, it is assumed that the repository is located at `https://github.com/vendor/package`.

The `packager:list` command now has an extra column showing the Git status of the package. If packages are version controlled, it will whether they're `Up to date`, `Ahead {number of commits}` or `Behind {number of commits}`. If not, they're displayed as `Not initialized`.

I've add a couple of tests to ensure the new features are working correctly. They both clone your `Jeroen-G/testassist` repository, but it doesn't really matter which repository they use, if you want to change it to something else.

Let me know if anything needs to be changed.

/Morten

mortenscheel commented:
Sorry about that. I'll take a look at the failing test case. I'll probably have it fixed later this weekend.

mortenscheel commented:
Okay, looks like the tests are green again.
Note that I've added `"cache_skeleton" => false` to the config file, which allows for faster installs (and tests).

Regarding style rules, I've tried a few different settings and plugins for PhpStorm, but nothing seems to match the StyleCI laravel preset. Do you have any tips you want to share?

Jeroen-G commented:
Perhaps you can also use the skeleton for the tests, or perhaps a proper composer.json is required?

Regarding styleci, I use their standard Laravel preset as far as I know: https://styleci.readme.io/docs/presets

mortenscheel commented:
Yeah, I've just changed the `packager:git` test use packager-skeleton. I couldn't make it work earlier, but now it does.

So... quite a lot of changes. I've tried my best to make the tests run faster, but I can't speed up the `composer require/remove` calls. I've probably even made them a bit slower, by adding `optimize-autoloader => true` to Testbench's `composer.json`. Without that change, I couldn't be 100% sure that the installed ServiceProvider was registered correctly in the autoload files.

I did manage to save some time, by not repeating the entire `packager:new` installation multiple times. After the first test, the resulting package (and the autoload-files created by Composer) are stored and reused later, to quickly install a "fake" package.

The "cached skeleton" feature is not used in the tests any longer, since the other way is much faster. So you're welcome to remove the feature if you like, but I figure maybe some users would want that option.

I've also refactored a lot of code, so most of my changes have ended up in traits. The intention was to make it less messy, but I'm open to your input if you want it structured differently.

Besides that, the biggest change is probably in the way I test whether packages packages are installed/removed correctly. I trust the tests a lot more now. And I've added a few extra assertions where I thought they made sense.

Anyhoo, let me know what you think.

Jeroen-G commented:
At this point, the speed of the tests don't really bother me. The fact that you trust the tests more now is much more appealing to me!

Jeroen-G commented on src/ComposerHandler.php:
```diff
@@ -0,0 +1,121 @@
+
+    /**
+     * Determines the path to Composer executable
+     * @todo Might not work on Windows
```
I have a Windows machine, so I will test this.

mortenscheel replied:
Cool. I've removed the @todo

Jeroen-G commented on tests/IntegratedTest.php:
```diff
@@ -3,38 +3,81 @@
     }
 
+    /**
+     * @depends test_new_package_is_created
```
I didn't know about `@depends`, looks very clever.

mortenscheel replied:
I don't normally like writing tests that rely on other tests, but sometimes it's super convenient.

Jeroen-G commented on config/packager.php:
```diff
@@ -7,6 +7,7 @@
      * Default: http://github.com/Jeroen-G/packager-skeleton/archive/master.zip
      */
     'skeleton' => 'http://github.com/Jeroen-G/packager-skeleton/archive/master.zip',
+    'cache_skeleton' => false,
```
What would be a proper place to document this? I'm thinking a line in just the config file might be enough?

mortenscheel replied:
Now I wonder if it's even worth keeping. The only benefit is that you save about a second of download time (if even that). On the other hands, it adds a layer of complexity, and people might end up using an outdated skeleton without knowing it.

Jeroen-G commented on src/ComposerHandler.php:
```diff
@@ -0,0 +1,121 @@
+     * @todo Might not work on Windows
+     * @return string
+     */
+    protected static function getComposerExecutable(): string
```
Why did you choose to make this static?

mortenscheel replied:
Hmm. Mostly just for convenience and old habits.
The whole initial setup of the Testbench environment runs in a static context (`setupBeforeClass`), before Laravel has even been initialized. This is necessary because I can't switch the app's `base_path` once it's already up and running.
But honestly I should change it back to non-static, since I can just call the methods from the `$instance` variable. I just didn't notice earlier :)

Jeroen-G commented on src/ProcessRunner.php:
```diff
@@ -0,0 +1,26 @@
+
+namespace JeroenG\Packager;
+
+trait ProcessRunner
```
Nice refactor 👍

mortenscheel replied:
Glad you like it. I considered calling the other one `ComposerCommander` or something like that, but decided to keep it boring...

mortenscheel commented:
WTF. One of the tests fail now, but only in PHP 7.2.
I'll have to install 7.2 and figure out why. I may not have time to fix it tonight though.

Jeroen-G commented:
One bug I found is that the repository is added to composer.json with a dash (i.e. vendor-package instead of vendor/package) and it is not added to the require array, so the composer requirements of the package are not installed.
Did you experience the same issues?

Jeroen-G commented:
Have you tried installing a package with tests? So far I can't get those tests to work :/

mortenscheel commented:
The failing unit test was caused by a simple timeout, which is fixed now.

> Have you tried installing a package with tests? So far I can't get those tests to work :/

Nope, I hadn't tried that until now, and I've been pulling my hair for a couple of hours without finding a good solution.
The root issue is that when you ´composer require´ a package, the package's dev-dependencies are ignored. Those are only installed if you run `composer install` in the package's root folder.

I tried using a [composer merge-plugin](https://github.com/wikimedia/composer-merge-plugin) which merges multiple composer.json files, but that only works if all the installed packages are compatible and don't have conflicting dependencies.

The way I see it, there's only one way to run package tests right now:
1. `cd projects/vendor/name`
2. `composer install`
3- `./vendor/bin/phpunit`

I tried writing a new Command to automate that workflow, but for some reason it won't work.

I'm afraid I'm all out of ideas here.
